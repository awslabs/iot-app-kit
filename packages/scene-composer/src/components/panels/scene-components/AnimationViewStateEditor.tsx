import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { FormField, Button, SpaceBetween, Autosuggest } from '@awsui/components-react';
import { useIntl } from 'react-intl';

import { IAnimationComponentInternal, useEditorState, ISceneNodeInternal } from '../../../store';
import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { animationObjectKey } from '../../three-fiber/AnimationComponent/AnimationComponent';

export type AnimationViewStateEditorProps = {
  node: ISceneNodeInternal;
  component: IAnimationComponentInternal;
  onUpdate: (currentAnimations: string[], noderef: string, component: IAnimationComponentInternal) => void;
};
export const AnimationViewStateEditor: React.FC<AnimationViewStateEditorProps> = ({
  node,
  component: animationComponent,
  onUpdate,
}: AnimationViewStateEditorProps) => {
  const { formatMessage } = useIntl();

  const sceneComposerId = useSceneComposerId();
  const { getObject3DBySceneNodeRef } = useEditorState(sceneComposerId);
  const object = getObject3DBySceneNodeRef(node.ref);
  const animationObject3D = object?.getObjectByName(animationObjectKey);
  const [selectedAnimations, setSelectedAnimations] = useState<string[] | undefined>();
  const animationPanelList = useRef<string[]>([]); // '' on end for next animation

  const animationOptions = useMemo(
    () =>
      animationObject3D?.animations.map((animation) => ({
        value: animation.name,
      })),
    [animationObject3D],
  );

  const updateAnimations = useCallback(
    (animation: string, i: number) => {
      if (selectedAnimations) {
        const currentAnimations = [...selectedAnimations];
        currentAnimations[i] = animation;
        setSelectedAnimations(currentAnimations);
        animationPanelList.current = [...currentAnimations, ''];
      }
    },
    [selectedAnimations],
  );

  const removeAnimation = useCallback(
    (i: number) => {
      if (selectedAnimations) {
        const currentAnimations = [...selectedAnimations];
        currentAnimations.splice(i, 1);
        setSelectedAnimations(currentAnimations);
        animationPanelList.current = [...currentAnimations, ''];
      }
    },
    [selectedAnimations],
  );

  useEffect(() => {
    const currentAnimations: string[] = [...animationComponent.currentAnimations] || [];
    setSelectedAnimations(currentAnimations);
    animationPanelList.current = [...currentAnimations, ''];
  }, [node.ref]);

  useEffect(() => {
    if (selectedAnimations && selectedAnimations != animationComponent.currentAnimations) {
      onUpdate(selectedAnimations, node.ref, animationComponent);
    }
  }, [selectedAnimations]);

  return (
    <SpaceBetween size='xs'>
      {animationPanelList.current.map((animation, i) => (
        <FormField
          key={`${i}-formField`}
          label={formatMessage({ defaultMessage: 'Choose Animations', description: 'FormField label' })}
        >
          {!!animation && (
            <Button
              key={`${i}-remove`}
              variant='link'
              data-testid={'removeButton' + i}
              onClick={() => removeAnimation(i)}
            >
              {formatMessage({
                defaultMessage: 'remove',
                description: 'label for the remove animation button',
              })}
            </Button>
          )}
          <Autosuggest
            key={`${i}-suggest`}
            onChange={({ detail }) => updateAnimations(detail.value, i)}
            value={animation}
            options={animationOptions}
            ariaLabel={formatMessage({
              defaultMessage: 'Autosuggest example with suggestions',
              description: 'Specifies the animation thats being selected',
            })}
            placeholder={formatMessage({
              defaultMessage: 'Enter Value',
              description: 'Default Message for before an animation is selected',
            })}
            empty={formatMessage({
              defaultMessage: 'No matches found',
              description: 'label for when the user searches for a nonexistent animation',
            })}
            enteredTextLabel={function (): string {
              return formatMessage({
                defaultMessage: 'No matches found',
                description: 'label for when the user searches for a nonexistent animation',
              });
            }}
          />
        </FormField>
      ))}
    </SpaceBetween>
  );
};
