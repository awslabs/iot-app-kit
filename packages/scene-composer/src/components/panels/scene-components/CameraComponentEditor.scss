.tm-camera-sub-form-label {
  padding: 10px 0px 0px 0px;
}
