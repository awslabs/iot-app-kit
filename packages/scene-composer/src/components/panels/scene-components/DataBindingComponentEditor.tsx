import { SpaceBetween } from '@awsui/components-react';
import React, { useCallback, useContext } from 'react';

import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import { useStore } from '../../../store';
import { IDataBindingComponentInternal } from '../../../store/internalInterfaces';
import { IComponentEditorProps } from '../ComponentEditor';

import { DataBindingMapEditor } from './common/DataBindingMapEditor';
import { Component } from '../../../models/SceneModels';

export interface IDataBindingComponentEditorProps extends IComponentEditorProps {
  component: IDataBindingComponentInternal;
}

export const DataBindingComponentEditor: React.FC<IDataBindingComponentEditorProps> = ({
  node,
  component,
}: IDataBindingComponentEditorProps) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const updateComponentInternal = useStore(sceneComposerId)((state) => state.updateComponentInternal);
  const removeComponent = useStore(sceneComposerId)((state) => state.removeComponent);
  const valueDataBindingProvider = useStore(sceneComposerId)(
    (state) => state.getEditorConfig().valueDataBindingProvider,
  );

  const onUpdateCallback = useCallback(
    (componentPartial: Partial<IDataBindingComponentInternal>, replace?: boolean) => {
      const componentPartialWithRef = { ref: component.ref, type: component.type, ...componentPartial };
      // When the data binding component has valueDataBindings left, update the component, otherwise remove
      // the whole component instead
      if (componentPartialWithRef.valueDataBinding) {
        updateComponentInternal(node.ref, componentPartialWithRef, replace);
      } else {
        removeComponent(node.ref, component.ref);
      }
    },
    [node.ref, component.ref],
  );

  return (
    <SpaceBetween size='s'>
      <DataBindingMapEditor
        allowPartialBinding
        skipFirstDivider
        hasBindingName={false}
        numFields={1}
        hasRemoveButton={false}
        valueDataBindingProvider={valueDataBindingProvider}
        component={{ ...component, valueDataBindings: [component.valueDataBinding as Component.IDataBindingMap] }}
        onUpdateCallback={onUpdateCallback}
      />
    </SpaceBetween>
  );
};
