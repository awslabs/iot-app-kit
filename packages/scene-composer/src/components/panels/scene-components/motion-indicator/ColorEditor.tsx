import React, { useEffect, useState, useContext } from 'react';
import { Box, Button, FormField, Grid, Icon, Select, SpaceBetween } from '@awsui/components-react';
import { isEmpty } from 'lodash';
import styled from 'styled-components';
import { defineMessages, useIntl } from 'react-intl';
import { ColorRepresentation } from 'three';

import { IMotionIndicatorComponentInternal, useStore } from '../../../../store';
import { sceneComposerIdContext } from '../../../../common/sceneComposerIdContext';
import { Component } from '../../../../models/SceneModels';
import { colors } from '../../../../utils/styleUtils';
import { Slider } from '../../Slider';
import { ColorPicker } from '../../ColorPicker/ColorPicker';

import { DataBindingEditor } from './DataBindingEditor';

const ColorSwatch = styled.div<{ backgroundColor: string }>`
  background-color: ${(props) => props.backgroundColor};
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: ${colors.containerBorderWhite};
  cursor: pointer;
`;

interface IColorEditorProps {
  component: IMotionIndicatorComponentInternal;
  selectedColorType:
    | Component.MotionIndicatorDataBindingName.BackgroundColor
    | Component.MotionIndicatorDataBindingName.ForegroundColor;
  onUpdateCallback: (componentPartial: any, replace?: boolean | undefined) => void;
}

const ColorEditor: React.FC<IColorEditorProps> = ({ component, selectedColorType, onUpdateCallback }) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const valueDataBindingProvider = useStore(sceneComposerId)(
    (state) => state.getEditorConfig().valueDataBindingProvider,
  );
  const { formatMessage } = useIntl();

  const [showColorPicker, setShowColorPicker] = useState(false);
  const [showColorDataBinding, setShowColorDataBinding] = useState(
    !isEmpty(component.valueDataBindings[selectedColorType]),
  );

  const toggleColorPicker = () => setShowColorPicker(!showColorPicker);
  const [showSlider, setShowSlider] = useState(false);
  const colorOptions = [
    { label: formatMessage({ defaultMessage: 'Select color', description: 'label' }), value: 'default' },
    { label: formatMessage({ defaultMessage: 'Add color rule', description: 'label' }), value: 'binding' },
  ];

  const i18ncolorEditorStrings = defineMessages({
    opacityButton: {
      defaultMessage: 'Opacity : {opacityPercentage, number, percent}',
      description: 'Button text for color opacity percentage',
    },
  });

  const selectedForegroundColor = selectedColorType === Component.MotionIndicatorDataBindingName.ForegroundColor;

  // hide slider when selected color type changes between foreground and background
  useEffect(() => {
    setShowSlider(false);
  }, [selectedColorType]);

  // reset states when component changed
  useEffect(() => {
    setShowColorDataBinding(!isEmpty(component.valueDataBindings[selectedColorType]));
    setShowColorPicker(false);
  }, [component.ref]);

  return (
    <SpaceBetween size='s'>
      <FormField
        label={
          selectedColorType === Component.MotionIndicatorDataBindingName.ForegroundColor
            ? formatMessage({ defaultMessage: 'Define arrow color', description: 'Select Color type option' })
            : formatMessage({ defaultMessage: 'Define background color', description: 'Select Color type option' })
        }
      >
        <Select
          data-testid='motion-indicator-color-select'
          selectedOption={showColorDataBinding ? colorOptions[1] : colorOptions[0]}
          onChange={(e) => {
            const value = e.detail.selectedOption.value;
            const useBinding = value === 'binding';
            let updatedComponent: IMotionIndicatorComponentInternal;
            if (!useBinding) {
              updatedComponent = {
                ...component,
                valueDataBindings: {
                  ...component.valueDataBindings,
                  [selectedColorType]: {},
                },
              };
            } else {
              updatedComponent = {
                ...component,
                config: {
                  ...component.config,
                  [selectedForegroundColor ? 'defaultForegroundColor' : 'defaultBackgroundColor']: undefined,
                },
              };
            }

            onUpdateCallback(updatedComponent, true);
            setShowColorDataBinding(useBinding);
            setShowColorPicker(false);
          }}
          options={colorOptions}
          selectedAriaLabel={formatMessage({
            defaultMessage: 'Selected',
            description:
              'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
          })}
          placeholder={formatMessage({ defaultMessage: 'Choose a color value', description: 'Placeholder' })}
        />
      </FormField>

      {selectedForegroundColor && !showColorDataBinding && (
        <FormField label={formatMessage({ defaultMessage: 'Arrow color', description: 'FormField label' })}>
          <Grid gridDefinition={[{ colspan: 2 }]}>
            <ColorSwatch
              data-testid='foreground-color-swatch'
              onClick={toggleColorPicker}
              backgroundColor={component.config.defaultForegroundColor as string}
            />
          </Grid>
        </FormField>
      )}

      {!selectedForegroundColor && (
        <FormField label={formatMessage({ defaultMessage: 'Background', description: 'FormField label' })}>
          <Grid gridDefinition={showColorDataBinding ? [{ colspan: 10 }] : [{ colspan: 2 }, { colspan: 10 }]}>
            {!showColorDataBinding && (
              <ColorSwatch
                data-testid='background-color-swatch'
                onClick={toggleColorPicker}
                backgroundColor={component.config.defaultBackgroundColor as string}
              />
            )}
            <Button data-testid='opacity-button' variant='normal' onClick={() => setShowSlider(!showSlider)}>
              {formatMessage(i18ncolorEditorStrings.opacityButton, {
                opacityPercentage: Math.round(component.config.backgroundColorOpacity * 100 ?? 100) / 100,
              })}
              <Box margin={{ left: 'xs' }} display='inline'>
                <Icon name={showSlider ? 'caret-down-filled' : 'caret-right-filled'} />
              </Box>
            </Button>
          </Grid>
          <Grid gridDefinition={showColorDataBinding ? [{ colspan: 10 }] : [{ colspan: 2 }, { colspan: 10 }]}>
            {!showColorDataBinding && <Box />}
            {showSlider && (
              <Slider
                value={component.config.backgroundColorOpacity * 100 ?? 100}
                step={1}
                min='0'
                max='100'
                onChange={(event) => {
                  const updatedComponent = {
                    ...component,
                    config: { ...component.config, backgroundColorOpacity: Number(event.target.value) / 100 },
                  };

                  onUpdateCallback(updatedComponent, true);
                }}
              />
            )}
          </Grid>
        </FormField>
      )}

      {showColorPicker && (
        <ColorPicker
          color={
            (selectedForegroundColor
              ? component.config.defaultForegroundColor
              : component.config.defaultBackgroundColor) as string
          }
          onChange={(newColor: ColorRepresentation) => {
            const updatedComponent = {
              ...component,
              config: {
                ...component.config,
                defaultBackgroundColor: selectedForegroundColor ? undefined : newColor,
                defaultForegroundColor: selectedForegroundColor ? newColor : undefined,
              },
            };

            onUpdateCallback(updatedComponent, true);
          }}
        />
      )}

      {showColorDataBinding && (
        <>
          <Box fontWeight='bold' padding={{ top: 's' }}>
            {formatMessage({ defaultMessage: 'Configure Rule', description: 'Box label' })}
          </Box>

          <DataBindingEditor
            valueDataBindingProvider={valueDataBindingProvider}
            component={component}
            onUpdateCallback={onUpdateCallback}
            dataBindingName={selectedColorType}
          />
        </>
      )}
    </SpaceBetween>
  );
};

ColorEditor.displayName = 'ColorEditor';

export default ColorEditor;
