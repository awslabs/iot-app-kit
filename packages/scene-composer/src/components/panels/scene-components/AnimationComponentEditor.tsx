import React, { useCallback, useContext } from 'react';

import { accessStore, ISceneNodeInternal, IAnimationComponentInternal } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';

import { AnimationViewStateEditor } from './AnimationViewStateEditor';

export type AnimationEditorProps = {
  node: ISceneNodeInternal;
  component: IAnimationComponentInternal;
};

export const AnimationComponentEditor: React.FC<AnimationEditorProps> = ({ ...props }) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const updateComponentInternal = accessStore(sceneComposerId)((state) => state.updateComponentInternal);
  // istanbul ignore next
  const onUpdate = useCallback(
    (currentAnimations, noderef, component) => {
      updateComponentInternal(noderef, { ...component, currentAnimations });
    },
    [updateComponentInternal],
  );

  return <AnimationViewStateEditor {...props} onUpdate={onUpdate} />;
};
