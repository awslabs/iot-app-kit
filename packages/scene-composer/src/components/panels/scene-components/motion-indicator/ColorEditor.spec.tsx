import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { mockComponent, mockProvider } from '../../../../../tests/components/panels/scene-components/MockComponents';
import { IMotionIndicatorComponentInternal, useStore } from '../../../../store';
import { KnownComponentType } from '../../../../interfaces';
import { Component } from '../../../../models/SceneModels';

import ColorEditor from './ColorEditor';

jest.mock('./DataBindingEditor', () => {
  const originalModule = jest.requireActual('./DataBindingEditor');
  return {
    ...originalModule,
    DataBindingEditor: (...props: any[]) => <div data-testid='DataBindingEditor'>{JSON.stringify(props)}</div>,
  };
});

let _sliderOnChangeCb;
jest.mock('../../Slider', () => {
  const originalModule = jest.requireActual('../../Slider');
  return {
    ...originalModule,
    Slider: (...props: any[]) => {
      _sliderOnChangeCb = props[0].onChange;
      return <div data-testid='Slider'>{JSON.stringify(props)}</div>;
    },
  };
});

jest.mock('../../ColorPicker/ColorPicker', () => ({
  ColorPicker: (props) => <div data-mocked='ColorPicker'>{JSON.stringify(props)}</div>,
}));

const baseState = {
  getEditorConfig: () => mockProvider,
} as any;

describe('ColorEditor', () => {
  const component: IMotionIndicatorComponentInternal = {
    ...mockComponent,
    type: KnownComponentType.MotionIndicator,
    shape: Component.MotionIndicatorShape.LinearPlane,
    valueDataBindings: {
      [Component.MotionIndicatorDataBindingName.BackgroundColor]: { ruleBasedMapId: 'rule-1' },
      [Component.MotionIndicatorDataBindingName.ForegroundColor]: { ruleBasedMapId: 'rule-2' },
    },
    config: {
      numOfRepeatInY: 4,
      backgroundColorOpacity: 0.5,
      defaultBackgroundColor: 'red',
      defaultForegroundColor: 'blue',
    },
  };
  const onUpdateCallback = jest.fn();

  it('should render correctly for background color with data binding', () => {
    useStore('default').setState(baseState);

    const { container } = render(
      <ColorEditor
        component={component}
        selectedColorType={Component.MotionIndicatorDataBindingName.BackgroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for background color without data binding', () => {
    useStore('default').setState(baseState);

    const rendered = render(
      <ColorEditor
        component={{ ...component, valueDataBindings: {} }}
        selectedColorType={Component.MotionIndicatorDataBindingName.BackgroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(rendered.container).toMatchSnapshot();

    // hide color picker when component.ref changed
    fireEvent.click(rendered.getByTestId('background-color-swatch'));
    fireEvent.click(rendered.getByTestId('opacity-button'));
    rendered.rerender(
      <ColorEditor
        component={{ ...component, valueDataBindings: {}, ref: 'new-ref' }}
        selectedColorType={Component.MotionIndicatorDataBindingName.BackgroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(rendered.container).toMatchSnapshot();
  });

  it('should render correctly for background color with slider and color picker shown', () => {
    useStore('default').setState(baseState);

    const rendered = render(
      <ColorEditor
        component={{ ...component, valueDataBindings: {} }}
        selectedColorType={Component.MotionIndicatorDataBindingName.BackgroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );

    fireEvent.click(rendered.getByTestId('background-color-swatch'));
    fireEvent.click(rendered.getByTestId('opacity-button'));

    expect(rendered.container).toMatchSnapshot();

    // hide sider when selected color type changed
    rendered.rerender(
      <ColorEditor
        component={{ ...component, valueDataBindings: {} }}
        selectedColorType={Component.MotionIndicatorDataBindingName.ForegroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(rendered.container).toMatchSnapshot();
  });

  it('should render correctly for foreground color with data binding', () => {
    useStore('default').setState(baseState);

    const { container } = render(
      <ColorEditor
        component={component}
        selectedColorType={Component.MotionIndicatorDataBindingName.ForegroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for foreground color without data binding', () => {
    useStore('default').setState(baseState);

    const { container } = render(
      <ColorEditor
        component={{ ...component, valueDataBindings: {} }}
        selectedColorType={Component.MotionIndicatorDataBindingName.ForegroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for foreground color with  color picker shown', () => {
    useStore('default').setState(baseState);

    const rendered = render(
      <ColorEditor
        component={{ ...component, valueDataBindings: {} }}
        selectedColorType={Component.MotionIndicatorDataBindingName.ForegroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );

    fireEvent.click(rendered.getByTestId('foreground-color-swatch'));

    expect(rendered.queryByTestId('opacity-button')).toBeFalsy();
    expect(rendered.container).toMatchSnapshot();

    // hide color picker when component.ref changed
    rendered.rerender(
      <ColorEditor
        component={{ ...component, valueDataBindings: {}, ref: 'new-ref' }}
        selectedColorType={Component.MotionIndicatorDataBindingName.ForegroundColor}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(rendered.container).toMatchSnapshot();
  });
});
