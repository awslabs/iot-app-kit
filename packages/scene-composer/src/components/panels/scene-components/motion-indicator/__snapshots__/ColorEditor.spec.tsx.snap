// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorEditor should render correctly for background color with data binding 1`] = `
<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define background color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Background"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":10}]"
      >
        <div
          data-mocked="Button"
          data-testid="opacity-button"
          variant="normal"
        >
          Opacity : 50%
          <div
            data-mocked="Box"
            display="inline"
            margin="{\\"left\\":\\"xs\\"}"
          >
            <div
              data-mocked="Icon"
              name="caret-right-filled"
            />
          </div>
        </div>
      </div>
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":10}]"
      />
    </div>
    <div
      data-mocked="Box"
      font-weight="bold"
      padding="{\\"top\\":\\"s\\"}"
    >
      Configure Rule
    </div>
    <div
      data-testid="DataBindingEditor"
    >
      [{"component":{"ref":"componentRef1","type":"MotionIndicator","uri":"componentUri","modelType":"componentModelType","shape":"LinearPlane","valueDataBindings":{"backgroundColor":{"ruleBasedMapId":"rule-1"},"foregroundColor":{"ruleBasedMapId":"rule-2"}},"config":{"numOfRepeatInY":4,"backgroundColorOpacity":0.5,"defaultBackgroundColor":"red","defaultForegroundColor":"blue"}},"dataBindingName":"backgroundColor"},{}]
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for background color with slider and color picker shown 1`] = `
.c0 {
  background-color: red;
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
  cursor: pointer;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define background color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Background"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2},{\\"colspan\\":10}]"
      >
        <div
          class="c0"
          data-testid="background-color-swatch"
        />
        <div
          data-mocked="Button"
          data-testid="opacity-button"
          variant="normal"
        >
          Opacity : 50%
          <div
            data-mocked="Box"
            display="inline"
            margin="{\\"left\\":\\"xs\\"}"
          >
            <div
              data-mocked="Icon"
              name="caret-down-filled"
            />
          </div>
        </div>
      </div>
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2},{\\"colspan\\":10}]"
      >
        <div
          data-mocked="Box"
        />
        <div
          data-testid="Slider"
        >
          [{"value":50,"step":1,"min":"0","max":"100"},{}]
        </div>
      </div>
    </div>
    <div
      data-testid="SketchPicker"
    >
      [{"disableAlpha":true,"color":"red"},{}]
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for background color with slider and color picker shown 2`] = `
.c0 {
  background-color: blue;
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
  cursor: pointer;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define arrow color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Arrow color"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2}]"
      >
        <div
          class="c0"
          data-testid="foreground-color-swatch"
        />
      </div>
    </div>
    <div
      data-testid="SketchPicker"
    >
      [{"disableAlpha":true,"color":"blue"},{}]
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for background color without data binding 1`] = `
.c0 {
  background-color: red;
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
  cursor: pointer;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define background color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Background"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2},{\\"colspan\\":10}]"
      >
        <div
          class="c0"
          data-testid="background-color-swatch"
        />
        <div
          data-mocked="Button"
          data-testid="opacity-button"
          variant="normal"
        >
          Opacity : 50%
          <div
            data-mocked="Box"
            display="inline"
            margin="{\\"left\\":\\"xs\\"}"
          >
            <div
              data-mocked="Icon"
              name="caret-right-filled"
            />
          </div>
        </div>
      </div>
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2},{\\"colspan\\":10}]"
      >
        <div
          data-mocked="Box"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for background color without data binding 2`] = `
.c0 {
  background-color: red;
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
  cursor: pointer;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define background color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Background"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2},{\\"colspan\\":10}]"
      >
        <div
          class="c0"
          data-testid="background-color-swatch"
        />
        <div
          data-mocked="Button"
          data-testid="opacity-button"
          variant="normal"
        >
          Opacity : 50%
          <div
            data-mocked="Box"
            display="inline"
            margin="{\\"left\\":\\"xs\\"}"
          >
            <div
              data-mocked="Icon"
              name="caret-down-filled"
            />
          </div>
        </div>
      </div>
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2},{\\"colspan\\":10}]"
      >
        <div
          data-mocked="Box"
        />
        <div
          data-testid="Slider"
        >
          [{"value":50,"step":1,"min":"0","max":"100"},{}]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for foreground color with  color picker shown 1`] = `
.c0 {
  background-color: blue;
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
  cursor: pointer;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define arrow color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Arrow color"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2}]"
      >
        <div
          class="c0"
          data-testid="foreground-color-swatch"
        />
      </div>
    </div>
    <div
      data-testid="SketchPicker"
    >
      [{"disableAlpha":true,"color":"blue"},{}]
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for foreground color with  color picker shown 2`] = `
.c0 {
  background-color: blue;
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
  cursor: pointer;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define arrow color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Arrow color"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2}]"
      >
        <div
          class="c0"
          data-testid="foreground-color-swatch"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for foreground color with data binding 1`] = `
<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define arrow color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}"
      />
    </div>
    <div
      data-mocked="Box"
      font-weight="bold"
      padding="{\\"top\\":\\"s\\"}"
    >
      Configure Rule
    </div>
    <div
      data-testid="DataBindingEditor"
    >
      [{"component":{"ref":"componentRef1","type":"MotionIndicator","uri":"componentUri","modelType":"componentModelType","shape":"LinearPlane","valueDataBindings":{"backgroundColor":{"ruleBasedMapId":"rule-1"},"foregroundColor":{"ruleBasedMapId":"rule-2"}},"config":{"numOfRepeatInY":4,"backgroundColorOpacity":0.5,"defaultBackgroundColor":"red","defaultForegroundColor":"blue"}},"dataBindingName":"foregroundColor"},{}]
    </div>
  </div>
</div>
`;

exports[`ColorEditor should render correctly for foreground color without data binding 1`] = `
.c0 {
  background-color: blue;
  width: 100%;
  height: 30px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
  cursor: pointer;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="FormField"
      label="Define arrow color"
    >
      <div
        data-mocked="Select"
        data-testid="motion-indicator-color-select"
        options="[{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"},{\\"label\\":\\"Add color rule\\",\\"value\\":\\"binding\\"}]"
        placeholder="Choose a color value"
        selectedarialabel="Selected"
        selectedoption="{\\"label\\":\\"Select color\\",\\"value\\":\\"default\\"}"
      />
    </div>
    <div
      data-mocked="FormField"
      label="Arrow color"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":2}]"
      >
        <div
          class="c0"
          data-testid="foreground-color-swatch"
        />
      </div>
    </div>
  </div>
</div>
`;
